Added parser for 'show access-list' and 'show access-list {id} to ASA

MikeJames15 opened the pull request:
## Description
<!--- Describe your changes in detail -->

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->

## Impact (If any)
<!--- If there is any negative impact - what is it? -->

## Screenshots:
<!--- Provide screenshots of tests/compile/demo/etc -->

## Checklist:
<!--- This is meant more as a personal checklist so we don't forgot important steps! -->
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
- [x] I have updated the changelog.
- [x] I have updated the documentation (If applicable).
- [x] I have added tests to cover my changes (If applicable).
- [x] All new and existing tests passed.
- [x] All new code passed compilation.

MikeJames15 commented:
This is my first time contributing. Please let me know if I messed something up. I tried to follow all the instructions.
I had trouble running the tests. Couldn't seem to be able to install cisco-distutils.

danielgraziano commented:
> This is my first time contributing. Please let me know if I messed something up. I tried to follow all the instructions. I had trouble running the tests. Couldn't seem to be able to install cisco-distutils.

No worries, the github checks run the tests and they seem to have passed

danielgraziano commented on changelog/undistributed/changelog_show_access_list_20210614202815.rst:
```diff
@@ -0,0 +1,8 @@
+--------------------------------------------------------------------------------
+* ASA
+    * Added ShowAccessList:
+      * show access-list
```
Please make sure the indentation is 4 spaces per indent. During the release process we use an automated script to generate the final changelog

danielgraziano commented on src/genie/libs/parser/asa/show_access_list.py:
```diff
@@ -0,0 +1,187 @@
+    	'access-list': {
+            Any(): {
+                Optional('elements'): int,
+#               'name_hash': str,
```
please remove if not needed

danielgraziano commented on src/genie/libs/parser/asa/show_access_list.py:
```diff
@@ -0,0 +1,187 @@
+    cli_command = 'show access-list'
+
+    def cli(self, output=None):
+        if output is None:
+            # excute command to get output
+            out = self.device.execute(self.cli_command)
+        else:
+            out = output
```
``` python
if output is None:
    output = self.device.execute(self.cli_command)

(snip)

for line in output.splitlines():

danielgraziano commented on src/genie/libs/parser/asa/show_access_list.py:
```diff
@@ -0,0 +1,187 @@
+            out = output
+
+        ret_dict = {}
+        ret_dict['access-list'] = {}
```
Please remove this. We only want to add to the dictionary within the for-loop. This way if nothing is parsed an empty dictionary will get returned (we expect this behaviour)

danielgraziano commented on src/genie/libs/parser/asa/show_access_list.py:
```diff
@@ -0,0 +1,187 @@
+                acl_name = groups['name']
+                ret_dict['access-list'][acl_name] = {}
+                ret_dict['access-list'][acl_name]['elements'] = int(groups['elements'])
+#                ret_dict['access-list'][acl_name]['name_hash'] = groups['namehash']
```
please remove if not used

danielgraziano replied:
But im curious. How come you are not parsing this?

MikeJames15 replied:
Added the value to the output

danielgraziano commented on src/genie/libs/parser/asa/show_access_list.py:
```diff
@@ -0,0 +1,187 @@
+            if m:
+                groups = m.groupdict()
+                acl_name = groups['name']
+                ret_dict['access-list'][acl_name] = {}
```
``` python
acl_dict = ret_dict.setdefault('access-list', {}).setdefault(acl_name, {})

danielgraziano commented on src/genie/libs/parser/asa/show_access_list.py:
```diff
@@ -0,0 +1,187 @@
+                groups = m.groupdict()
+                acl_name = groups['name']
+                ret_dict['access-list'][acl_name] = {}
+                ret_dict['access-list'][acl_name]['elements'] = int(groups['elements'])
```
``` python
acl_dict.update({'elements': int(groups['elements']})

danielgraziano commented on src/genie/libs/parser/asa/show_access_list.py:
```diff
@@ -0,0 +1,187 @@
+                groups = m.groupdict()
+                acl_name = groups['name']
+                line_num = groups['entry']
+                if not acl_name in ret_dict['access-list'].keys() :
+                    ret_dict['access-list'][acl_name] = {}
+                if not 'entry' in ret_dict['access-list'][acl_name].keys():
+                    ret_dict['access-list'][acl_name]['entry'] = {}
```
The magic of `<dict>.setdefault()` is it wont overwrite existing keys. It will only insert if the key doesnt exist.
``` python
acl_entry_dict = ret_dict.setdefault('access-list', {}).setdefault(acl_name, {}).setdefault('entry', {})

danielgraziano commented on src/genie/libs/parser/asa/show_access_list.py:
```diff
@@ -0,0 +1,187 @@
+            if m:
+                groups = m.groupdict()
+                line_num = groups['entry']
+                if not acl_name in ret_dict['access-list'].keys() :
+                    ret_dict['access-list'][acl_name] = {}
+                if not 'entry' in ret_dict['access-list'][acl_name].keys():
+                    ret_dict['access-list'][acl_name]['entry'] = {}
```
same with setdefault

danielgraziano commented on src/genie/libs/parser/asa/show_access_list.py:
```diff
@@ -0,0 +1,187 @@
+            if m:
+                groups = m.groupdict()
+                ret_dict['alert_interval'] = int(groups['alert_interval'])
+#            else:
+#                print('Not matching line: {}'.format(line))
```
please remove

danielgraziano commented on src/genie/libs/parser/asa/show_access_list.py:
```diff
@@ -0,0 +1,187 @@
+#                print('Not matching line: {}'.format(line))
+
+
+        if not(bool(ret_dict['access-list'])):
+            ret_dict = {}
```
please remove, not needed if the other comments are updated

danielgraziano commented on src/genie/libs/parser/asa/show_access_list.py:
```diff
@@ -0,0 +1,187 @@
+            ret_dict = {}
+        return ret_dict
+
+    def add_if_in_groups(self, line_dict, groups, names):
+        for name in names:
+            if name in groups and not groups[name] is None:
+                line_dict[name] = groups[name]
+
+    def add_host(self, line_dict, groups, prefix):
+        names = ['host', 'any', 'object_group']
+        for name in names:
+            prefixed_name = prefix + name
+            if prefixed_name in groups and not groups[prefixed_name] is None:
+                line_dict[name] = groups[prefixed_name]
```
Please do this logic within each regex match. There shouldnt be any more methods defined so that all parsers follow the same format

danielgraziano commented on src/genie/libs/parser/asa/show_access_list.py:
```diff
@@ -0,0 +1,309 @@
+# Python
+import re
+import logging
+from collections import OrderedDict
```
No need to use an OrderedDict, please use a regular dictionary

MikeJames15 replied:
In access lists it makes a difference in which order the entries appear, for example when looping over them it is not the same if we use a regular dict or an OrderedDict. An OrderedDict will provide them in the right order and a regular dict will not guarantee it.

danielgraziano commented on src/genie/libs/parser/asa/show_access_list.py:
```diff
@@ -0,0 +1,309 @@
+
+# Python
+import re
+import logging
```
Not used, please remove

MikeJames15 replied:
I have removed the logging import

danielgraziano commented on src/genie/libs/parser/asa/show_access_list.py:
```diff
@@ -0,0 +1,309 @@
+            if line == '':
+                continue
+
+            m = p1.match(line)
```
please add a comment above each regex usage that shows what lines the regex can match (same ones from above). Please do below as well

MikeJames15 replied:
Copied the comments from the pattern definitions to the match lines such as "m = p1.match(line)"

danielgraziano reviewed (changes requested):
Should be good after these changes

MikeJames15 commented:
My tests have started to fail without making any changes ( before I did my last commit  and without any changes from when they were working). What should I do?

MikeJames15 commented:
I have some changes to show_bgp.py for iosxe. 

The changes include:

- adding up_time, down_time, last_read and last_write to the neighbor level. these values were appearing under address_family in the schema for some reason even though they are not dependent on the address_family and they were not getting set.

- fixed bug where split lines were not processed. When a line ends after the prefix, it is saved and joined with the following one.

- added support for AS with a dot in them (65324.3) in the AS path. Only managed to do this in one of the regexes (p6) the other one was more complicated to implement. May do it in the future.

- used setdefault() instead of an if followed by creation of dicts, as recommended for show_access_list.py.

- removed empty address_family entry created when only data for a RD is available.

Have not pushed this change yet. Should I add it to this pull request, wait for this to be resolved, add it to another branch in my fork or do a separate fork for it?

Please let me know how to continue.
